File: src/python/explog.py
import libpinocchio_pywrap as se3
import numpy as np
import numpy.linalg as npl
from math import sin,cos
import math
from utils import skew,cross,eye

def exp3(v):
    '''
    Exp: so3 -> SO3. Return the integral of the input angular velocity during time 1.
    '''
    nv = npl.norm(v)
    return se3.AngleAxis(nv,v/nv).matrix()

def log3(R):
    '''
    Log: SO3 -> so3. Pseudo-inverse of log from SO3 -> { v \in so3, ||v|| < 2pi }.
    '''
    return se3.AngleAxis(R).vector()

def exp6( nu ):
    '''
    Exp: se3 -> SE3. Return the integral of the input spatial velocity during time 1.
    '''
    if isinstance(nu,se3.Motion): w = nu.angular; v = nu.linear
    else:    w = nu[3:]; v = nu[:3]
    if npl.norm(w)>1e-15:
        R = exp3(w)
        t = npl.norm(w)
        S = skew(w)
        V = eye(3) + (1-cos(t))/t**2 * S + (t-sin(t))/t**3 * S*S
        p = V*v
        return se3.SE3(R,p)
    else:
        return se3.SE3(eye(3),v)

def log6( m ):
    '''
    Log: SE3 -> se3. Pseudo-inverse of exp from SE3 -> { v,w \in se3, ||w|| < 2pi }.
    '''
    if isinstance(m,se3.SE3): R = m.rotation; p = m.translation
    else: R = m[:3,:3]; p = m[:3,3]
    w = log3(R)
    if npl.norm(w)>1e-15:
        t = npl.norm(w)
        S = skew(w)
        V = eye(3) + (1-cos(t))/t**2 * S + (t-sin(t))/t**3 * S*S
        v = npl.inv(V)*p
    else:
        v = p
    return se3.Motion(v,w)

def exp(x):
    if isinstance(x,se3.Motion): return exp6(x)
    elif np.isscalar(x): return math.exp(x)
    elif isinstance(x,np.matrix):
        if len(x)==6: return exp6(x)
        elif len(x)==3: return exp3(x)
        else: print 'Error only 3 and 6 vectors are allowed.'
    else: print 'Error exp is only defined for real, vector3, vector6 and se3.Motion objects.'

def log(x):
    if isinstance(x,se3.SE3): return log6(x)
    elif np.isscalar(x): return math.log(x)
    elif isinstance(x,np.matrix):
        if len(x)==6: return log6(x)
        elif len(x)==3: return log3(x)
        else: print 'Error only 3 and 6 vectors are allowed.'
    else: print 'Error log is only defined for real, vector3, vector6 and se3.SE3 objects.'


__all__ = [ 'exp3', 'log3', 'exp6', 'log6','exp','log' ]

File: bindings/nanobind/spatial/explog.cpp
// Copyright (c) 2026 INRIA

#include "pinocchio/bindings/python-nb/fwd.hpp"

#include "pinocchio/spatial.hpp"

#include <nanobind/eigen/dense.h>

PINOCCHIO_PYTHON_NAMESPACE_BEGIN
using namespace nb::literals;

using Vector3 = Eigen::Matrix<Scalar, 3, 1, Options>;
using Vector4 = Eigen::Matrix<Scalar, 4, 1, Options>;
using Vector6 = Eigen::Matrix<Scalar, 6, 1, Options>;
using Vector7 = Eigen::Matrix<Scalar, 7, 1, Options>;
using Matrix3 = Eigen::Matrix<Scalar, 3, 3, Options>;
using Matrix4 = Eigen::Matrix<Scalar, 4, 4, Options>;
using Matrix6 = Eigen::Matrix<Scalar, 6, 6, Options>;
using Quaternion = Eigen::Quaternion<Scalar, Options>;
using ConstQuaternionMap = Eigen::Map<const Quaternion>;

void exposeExplog(nb::module_ m)
{
  // --- SO3 exponential map ---

  m.def(
    "exp3", [](const Vector3 & v) -> Matrix3 { return pinocchio::exp3(v); }, "w"_a,
    "Exp: so3 -> SO3. Return the integral of the input"
    " vector w during time 1. This is also known as the Rodrigues formula.");

  m.def(
    "exp3_quat",
    [](const Vector3 & v) -> Vector4 { return pinocchio::quaternion::exp3(v).coeffs(); }, "w"_a,
    "Exp: so3 -> S3. Returns the integral of the input vector w during time 1, represented "
    "as a unit Quaternion.");

  m.def(
    "Jexp3",
    [](const Vector3 & v) -> Matrix3 {
      Matrix3 res;
      pinocchio::Jexp3(v, res);
      return res;
    },
    "w"_a,
    "Jacobian of exp(v) which maps from the tangent of SO(3) at R = exp(v) to"
    " the tangent of SO(3) at Identity.");

  // --- SO3 logarithmic map ---

  m.def(
    "log3", [](const Matrix3 & R) -> Vector3 { return pinocchio::log3(R); }, "R"_a,
    "Log: SO3 -> so3 is the pseudo-inverse of Exp: so3 -> SO3. Log maps from SO3"
    " -> { v in so3, ||v|| < 2pi }.");

  m.def(
    "log3",
    [](const Vector4 & quat_vec) -> Vector3 {
      ConstQuaternionMap q(quat_vec.data());
      return pinocchio::quaternion::log3(q);
    },
    "quat"_a,
    "Log: S^3 -> so3 is the pseudo-inverse of Exp: so3 -> S^3, the exponential map from so3 to "
    "the unit quaternions. It maps from S^3 -> { v in so3, ||v|| < 2pi }.");

  m.def(
    "Jlog3",
    [](const Matrix3 & R) -> Matrix3 {
      Matrix3 res;
      pinocchio::Jlog3(R, res);
      return res;
    },
    "R"_a,
    "Jacobian of log(R) which maps from the tangent of SO(3) at R to"
    " the tangent of SO(3) at Identity.");

  m.def(
    "Hlog3",
    [](const Matrix3 & R, const Vector3 & v) -> Matrix3 {
      Matrix3 res;
      pinocchio::Hlog3(R, v, res);
      return res;
    },
    "R"_a, "v"_a, "v^T * H where H is the Hessian of log(R)");

  // --- SE3 exponential map ---

  m.def(
    "exp6", [](const pinocchio::Motion & v) -> pinocchio::SE3 { return pinocchio::exp6(v); },
    "motion"_a,
    "Exp: se3 -> SE3. Return the integral of the input"
    " spatial velocity during time 1.");

  m.def(
    "exp6", [](const Vector6 & v) -> pinocchio::SE3 { return pinocchio::exp6(v); }, "v"_a,
    "Exp: se3 -> SE3. Return the integral of the input"
    " spatial velocity during time 1.");

  m.def(
    "exp6_quat", [](const Vector6 & v) -> Vector7 { return pinocchio::quaternion::exp6(v); }, "v"_a,
    "Exp: se3 -> R3 * S3. Return the integral of the input 6D spatial velocity over unit time,"
    " using quaternion to represent rotation as in the standard configuration layout"
    " for the Lie group SE3.");

  m.def(
    "Jexp6",
    [](const pinocchio::Motion & v) -> Matrix6 {
      Matrix6 res;
      pinocchio::Jexp6(v, res);
      return res;
    },
    "motion"_a,
    "Jacobian of exp(v) which maps from the tangent of SE(3) at exp(v) to"
    " the tangent of SE(3) at Identity.");

  m.def(
    "Jexp6",
    [](const Vector6 & vec6) -> Matrix6 {
      pinocchio::MotionRef<const Vector6> v(vec6);
      Matrix6 res;
      pinocchio::Jexp6(v, res);
      return res;
    },
    "v"_a,
    "Jacobian of exp(v) which maps from the tangent of SE(3) at exp(v) to"
    " the tangent of SE(3) at Identity.");

  // --- SE3 logarithmic map ---

  m.def(
    "log6", [](const pinocchio::SE3 & M) -> pinocchio::Motion { return pinocchio::log6(M); }, "M"_a,
    "Log: SE3 -> se3. Pseudo-inverse of exp from SE3"
    " -> { v,w in se3, ||w|| < 2pi }.");

  m.def(
    "log6", [](const Matrix4 & H) -> pinocchio::Motion { return pinocchio::log6(H); },
    "homegeneous_matrix"_a,
    "Log: SE3 -> se3. Pseudo-inverse of Exp: so3 -> SO3. Log maps from SE3"
    " -> { v,w in se3, ||w|| < 2pi }.");

  m.def(
    "log6_quat",
    [](const Vector7 & q) -> pinocchio::Motion {
      ConstQuaternionMap quat(q.template tail<4>().data());
      const Vector3 v(q.template head<3>());
      return pinocchio::log6(quat, v);
    },
    "q"_a,
    "Log: R^3 * S^3 -> se3. Pseudo-inverse of Exp: se3 -> R^3 * S^3,"
    " the variant of the SE3 Exp using quaternions for the rotations.");

  m.def(
    "Jlog6", [](const pinocchio::SE3 & M) -> Matrix6 { return pinocchio::Jlog6(M); }, "M"_a,
    "Jacobian of log(M) which maps from the tangent of SE(3) at M to"
    " the tangent of SE(3) at Identity.");
}
PINOCCHIO_PYTHON_NAMESPACE_END
